File: si/lab03/AStar.java
import java.util.List;
import java.util.LinkedList;
import java.util.Map;
import java.util.HashMap;
import java.util.Iterator;
import java.io.File;
import java.io.FileReader;
import java.io.BufferedReader;
import java.io.IOException;

/**
 * Implementacja algorytmu A*.
 *
 * Wynik działania algorytmu: najkrótsza ścieżka w grafie
 * łącząca węzeł początkowy z węzłem końcowym.
 *
 * Użycie: java AStar plik_grafu plik_heurezy nazwa_węzła_początkowego
 *
 * Za węzeł końcowy uznawany jest węzeł o heurezie równej 0.
 *
 * @author Michał Dettlaff
 */

public class AStar {
	/**
	 * Graf skierowany przechowujący miasta, odległości między nimi
	 * i heurystykę. Kluczami są nazwy miast.
	 */
	private static Map<String, Node> graph = new HashMap<String, Node>();
	/**
	 * Węzeł początkowy.
	 */
	private static Node w0;
	/**
	 * Węzeł końcowy.
	 */
	private static Node wg;
	/**
	 * Lista węzłów do rozwinięcia.
	 */
	private static List<Node> open = new LinkedList<Node>();
	/**
	 * Lista zbadanych węzłów.
	 */
	private static List<Node> closed = new LinkedList<Node>();

	private AStar() {};

	public static void main(String[] args) {
		if (args.length > 2) {
			try {
				LinkedList<Node> path;

				readInputData(args);

				path = AStar(w0, wg);

				if (path != null) {
					System.out.println("Najkrótsza ścieżka od " +
							w0.getName() + " do " + wg.getName() + ":");
					System.out.print(w0.getName());
					Iterator<Node> iterator = path.descendingIterator();
					while (iterator.hasNext()) {
						System.out.print(" -> " + iterator.next().getName());
					}
					System.out.println();
				} else {
					System.out.println("Nie można znaleźć ścieżki.");
				}
			} catch (IOException e) {
				System.out.println("Nie można odczytać podanego pliku.");
				System.exit(1);
			}
		} else {
			System.out.println(
					"Implementacja algorytmu A*.\n" +
					"Wynik działania algorytmu: najkrótsza ścieżka w grafie\n" +
					"łącząca węzeł początkowy z węzłem końcowym.\n\n" +
					"Użycie: java AStar plik_grafu plik_heurezy " +
					"nazwa_węzła_początkowego\n" +
					"Za węzeł końcowy uznawany jest węzeł o heurezie równej 0."
					);
		}
	}

	/**
	 * Algorytm A*. Znajduje najkrótszą ścieżkę od węzła w0 do węzła wg.
	 *
	 * @return Najkrótsza ścieżka pomiędzy podanymi węzłami. Jeśli nie
	 *         znaleziono ścieżki, zwraca null.
	 */
	public static LinkedList<Node> AStar(Node w0, Node wg) {
		open.add(w0);
		w0.setG(0.0);
		w0.setF(w0.getH());
		while (!open.isEmpty()) {
			// z listy open wybierz węzeł w o najniższym koszcie f(w)
			Node node = minimum(open);
			if (node.equals(wg)) {
				return node.path();
			}
			open.remove(node);
			closed.add(node);
			/*
			 * oceń sąsiadów węzła w, tzn. dla każdego v ∈ N(w) wyznacz
			 * g(v) = g(w) + κ(w, v), oblicz h(v) oraz f(v) = g(v) + h(v);
			 * dodaj do listy open wszystkich nieodwiedzonych sąsiadów
			 * węzła w, którzy albo nie występują w open, albo są już w open,
			 * ale przypisany im koszt jest większy od kosztu aktualnie
			 * wyznaczonego
			 */
			for (String neighborName : node.getNeighbors().keySet()) {
				Node neighbor = graph.get(neighborName);
				if (closed.contains(neighbor)) {
					continue;
				}
				Double currentG = node.getG() +
					node.getNeighbors().get(neighbor.getName());
				boolean currentIsBetter = false;
				if (!open.contains(neighbor)) {
					open.add(neighbor);
					currentIsBetter = true;
				} else if (currentG < neighbor.getG()) {
					currentIsBetter = true;
				}
				if (currentIsBetter) {
					neighbor.setPrevious(node);
					neighbor.setG(currentG);
					neighbor.setF(neighbor.getG() + neighbor.getH());
				}
			}
		}
		return null;
	}

	/**
	 * Zwraca węzeł o najmniejszym koszcie z listy.
	 */
	private static Node minimum(List<Node> nodes) {
		Node minimum = nodes.get(0);
		for (Node node : nodes) {
			if (node.getF() < minimum.getF()) {
				minimum = node;
			}
		}
		return minimum;
	}

	private static void readInputData(String[] args) throws IOException {
		BufferedReader reader = new BufferedReader(new FileReader(
					args[0]));
		String inputLine;

		while ((inputLine = reader.readLine()) != null) {
			String[] nodeInfo = inputLine.split(" ");
			if (!graph.containsKey(nodeInfo[0])) {
				Node node = new Node(nodeInfo[0]);
				node.getNeighbors().put(nodeInfo[1], 
						Double.parseDouble(nodeInfo[2]));
				graph.put(nodeInfo[0], node);
			} else {
				graph.get(nodeInfo[0]).getNeighbors().put(nodeInfo[1], 
						Double.parseDouble(nodeInfo[2]));
			}
		}

		reader = new BufferedReader(new FileReader(args[1]));
		while ((inputLine = reader.readLine()) != null) {
			String[] nodeInfo = inputLine.split(" ");
			if (Integer.parseInt(nodeInfo[1]) == 0) {
				wg = graph.get(nodeInfo[0]);
			}
			if (!graph.containsKey(nodeInfo[0])) {
				graph.put(nodeInfo[0], new Node(nodeInfo[0]));
			}
			graph.get(nodeInfo[0]).setH(
					Double.parseDouble(nodeInfo[1]));
		}

		w0 = graph.get(args[2]);
		if (!graph.containsValue(w0)) {
			System.out.println("Podany węzeł początkowy nie istnieje");
			System.exit(1);
		}
	}
}


/**
 * Węzeł w grafie.
 */
class Node {
	/**
	 * Nazwa identyfikująca węzeł (nazwa miasta).
	 */
	private String name;
	/**
	 * Sąsiedzi (ich nazwy) i przypisane im koszty.
	 */
	private Map<String, Double> neighbors = new HashMap<String, Double>();
	/**
	 * Skąd przyszliśmy do tego węzła.
	 */
	private Node previous;
	/**
	 * Heurystyka.
	 */
	private Double h;
	/**
	 * Koszt ścieżki od w0 do tego węzła.
	 */
	private Double g;
	/**
	 * Funkcja kosztu.
	 */
	private Double f;

	public Node(String name) {
		this.name = name;
	}

	public String getName() {
		return name;
	}

	public Map<String, Double> getNeighbors() {
		return neighbors;
	}
	public void setNeighbors(Map<String, Double> neighbors) {
		this.neighbors = neighbors;
	}

	public Node getPrevious() {
		return previous;
	}
	public void setPrevious(Node previous) {
		this.previous = previous;
	}

	public Double getH() {
		return h;
	}
	public void setH(Double h) {
		this.h = h;
	}

	public Double getG() {
		return g;
	}
	public void setG(Double g) {
		this.g = g;
	}

	public Double getF() {
		return f;
	}
	public void setF(Double f) {
		this.f = f;
	}


	public String toString() {
		String previousName;
		if (previous != null) {
			previousName = previous.getName();
		} else {
			previousName = null;
		}
		return name + ": " + neighbors + ", prev: " + previousName +
			", h: " + h;
	}

	/**
	 * Zwraca ścieżkę od tego węzła do w0.
	 */
	public LinkedList<Node> path() {
		LinkedList<Node> path = new LinkedList<Node>();
		Node w = this;
		while (w.getPrevious() != null) { // previous == null dla w0
			path.add(w);
			w = w.getPrevious();
		}
		return path;
	}

	public boolean equals(Object obj) {
		Node node = (Node) obj;
		return this.name.equals(node.getName());
	}
}

File: si/lab03/astar.py
#!/usr/bin/env python
# -*- coding: utf8 -*-

import sys

usage = '''\
Algorytm A*, znajdujący najkrótszą ścieżkę pomiędzy danymi wierzchołkami grafu.
Użycie: python astar.py PLIK_GRAFU PLIK_HEUREZY WĘZEŁ_POCZĄTKOWY\
'''
# węzeł końcowy to węzeł o heurezie równiej 0

graph = {} # klucz: nazwa węzła, wartość: węzeł
open_nodes = [] # węzły do rozwinięcia
closed_nodes = [] # zbadane węzły
w0 = None
wg = None

class Node:
	def __init__(self, name):
		self.name = name
		self.neighbors = {} # klucz: węzeł sąsiadujący, wartość: koszt
		self.previous = None

	def __cmp__(self, node):
		return cmp(self.f, node.f)

	def __eq__(self, node):
		return self.name == node.name

	def __hash__(self):
		return hash(self.name)

	def __str__(self):
		description = self.name + ' ->'
		for neighbor, cost in self.neighbors.iteritems():
			description += ' ' + neighbor.name + ' ' + str(cost)
		description += ', h=' + str(self.h)
		if self.previous:
			description += ', prev=' + self.previous.name
		return description


def read_input():
	global w0
	global wg
	graph_filename = sys.argv[1]
	h_filename = sys.argv[2]
	w0_name = sys.argv[3]
	# tworzymy węzły
	graph_file = open(graph_filename)
	for line in graph_file.readlines():
		line_data = line.split(' ')
		node_name = line_data[0]
		node = Node(node_name)
		if not node_name in graph:
			graph[node_name] = Node(node_name)
	graph_file.close()
	# dodajemy informacje o sąsiadach (krawędziach) do węzłów
	graph_file = open(graph_filename)
	for line in graph_file.readlines():
		line_data = line.split(' ')
		node_name = line_data[0]
		neighbor_name = line_data[1]
		neighbor = graph[neighbor_name]
		cost = float(line_data[2].strip())
		graph[node_name].neighbors[neighbor] = cost
	graph_file.close()
	# dodajemy informacje o heurezie do węzłów
	h_file = open(h_filename)
	for line in h_file.readlines():
		line_data = line.split(' ')
		node_name = line_data[0]
		h = float(line_data[1].strip())
		graph[node_name].h = h
		if h == 0:
			wg = graph[node_name]
	h_file.close()
	if w0_name not in graph:
		print 'Podanego węzła początkowego nie ma w grafie.'
		sys.exit(1)
	w0 = graph[w0_name]

def a_star(w0, wg):
	'Algorytm A*.'
	open_nodes.append(w0)
	w0.g = .0
	w0.f = w0.h
	while len(open_nodes) != 0:
		node = min(open_nodes)
		if node == wg:
			return path(wg)
		open_nodes.remove(node)
		closed_nodes.append(node)
		for neighbor in node.neighbors:
			if neighbor in closed_nodes:
				continue
			current_g = node.g + node.neighbors[neighbor]
			is_current_better = False
			if neighbor not in open_nodes:
				open_nodes.append(neighbor)
				is_current_better = True
			elif current_g < neighbor.g:
				is_current_better = True
			if is_current_better:
				neighbor.previous = node
				neighbor.g = current_g
				neighbor.f = neighbor.g + neighbor.h
	return None

def path(node):
	'Zwraca ścieżkę powstałą w wyniku działania algorytmu A*.'
	path = []
	while node.previous:
		path.append(node)
		node = node.previous
	return reversed(path)


if __name__ == '__main__':
	try:
		if len(sys.argv) > 3:
			read_input()
			print 'Graf wejściowy:'
			for node in graph.values():
				print node

			path = a_star(w0, wg)

			if path is not None:
				print '\nNajkrótsza ścieżka z', w0.name, 'do', wg.name + ':'
				path_length = .0
				print w0.name, path_length,
				for node in path:
					path_length += node.previous.neighbors[node]
					print '->', node.name, path_length,
			else:
				print '\nNie znaleziono ścieżki.'
		else:
			print usage
	except IOError:
		print 'Nie można odnaleźć podanego pliku.'
		sys.exit(1)

